File: app/src/main/java/com/liadpaz/amp/adapters/PlaylistAdapter.java
package com.liadpaz.amp.adapters;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import com.bumptech.glide.Glide;
import com.liadpaz.amp.databinding.ItemPlaylistSongBinding;
import com.liadpaz.amp.databinding.ItemSongShuffleBinding;
import com.liadpaz.amp.interfaces.ItemTouchHelperAdapter;
import com.liadpaz.amp.interfaces.OnRecyclerItemClickListener;
import com.liadpaz.amp.interfaces.OnStartDragListener;
import com.liadpaz.amp.livedatautils.PlaylistsUtil;
import com.liadpaz.amp.livedatautils.QueueUtil;
import com.liadpaz.amp.utils.Utilities;
import com.liadpaz.amp.viewmodels.Song;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PlaylistAdapter extends ListAdapter<Song, PlaylistAdapter.SongViewHolder> implements ItemTouchHelperAdapter {
    private static final String TAG = "AmpApp.PlaylistAdapter";
    private static final int TYPE_ITEM = 1;

    private ArrayList<Song> songs;

    private OnRecyclerItemClickListener onMoreClickListener;
    private View.OnClickListener onShuffleClickListener;
    private OnStartDragListener onStartDragListener;
    private ItemTouchHelperAdapter itemTouchHelperAdapter;

    private Context context;

    public PlaylistAdapter(@NonNull Context context, @NonNull OnRecyclerItemClickListener onMoreClickListener, @NonNull View.OnClickListener onShuffleClickListener, @NonNull ItemTouchHelperAdapter itemTouchHelperAdapter) {
        super(new DiffUtil.ItemCallback<Song>() {
            @Override
            public boolean areItemsTheSame(@NonNull Song oldItem, @NonNull Song newItem) { return oldItem == newItem; }

            @Override
            public boolean areContentsTheSame(@NonNull Song oldItem, @NonNull Song newItem) { return oldItem.equals(newItem); }
        });

        this.context = context;
        this.onMoreClickListener = onMoreClickListener;
        this.onShuffleClickListener = onShuffleClickListener;
        this.itemTouchHelperAdapter = itemTouchHelperAdapter;
    }

    public void setOnStartDragListener(@NonNull OnStartDragListener onStartDragListener) {
        this.onStartDragListener = onStartDragListener;
    }

    @NonNull
    @Override
    public SongViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == TYPE_ITEM) {
            return new SongViewHolder(ItemPlaylistSongBinding.inflate(LayoutInflater.from(context), parent, false), (v, position) -> {
                QueueUtil.setQueue(new ArrayList<>(songs));
                QueueUtil.setPosition(position - 1);
            }, onShuffleClickListener);
        }
        return new SongViewHolder(ItemSongShuffleBinding.inflate(LayoutInflater.from(context), parent, false), (v, position) -> {}, onShuffleClickListener);
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onBindViewHolder(@NonNull SongViewHolder holder, int position) {
        if (position != 0) {
            Song song = getItem(position - 1);

            ItemPlaylistSongBinding binding = (ItemPlaylistSongBinding)holder.binding;

            binding.tvSongName.setText(song.songTitle);
            binding.tvSongArtist.setText(Utilities.joinArtists(song.songArtists));

            Glide.with(context).load(Utilities.getCoverUri(song)).into(binding.ivSongCover);

            binding.btnMore.setOnClickListener(v -> onMoreClickListener.onItemClick(v, position - 1));
            binding.btnDrag.setOnTouchListener((v, event) -> {
                if (event.getAction() == MotionEvent.ACTION_DOWN) {
                    onStartDragListener.onStartDrag(holder);
                }
                return true;
            });
        }
    }

    @Override
    public void onItemMove(int fromPosition, int toPosition) {
        PlaylistsUtil.setIsChanging(true);
        Collections.swap(songs, fromPosition - 1, toPosition - 1);
        itemTouchHelperAdapter.onItemMove(fromPosition - 1, toPosition - 1);
        notifyItemMoved(fromPosition, toPosition);
    }

    @Override
    public void onItemDismiss(int position) {
        throw new UnsupportedOperationException();
    }

    @Override
    public int getItemViewType(int position) {
        return position == 0 ? 0 : TYPE_ITEM;
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + 1;
    }

    @Override
    public void submitList(@Nullable List<Song> list) {
        super.submitList(songs = list != null ? new ArrayList<>(list) : null);
    }

    static class SongViewHolder extends RecyclerView.ViewHolder {
        private ViewBinding binding;

        SongViewHolder(@NonNull ViewBinding binding, @NonNull OnRecyclerItemClickListener onItemClick, @NonNull View.OnClickListener onShuffleClickListener) {
            super(binding.getRoot());
            this.binding = binding;

            if (binding instanceof ItemSongShuffleBinding) {
                ((ItemSongShuffleBinding)binding).getRoot().setOnClickListener(onShuffleClickListener);
            } else {
                itemView.setOnClickListener(v -> onItemClick.onItemClick(v, getAdapterPosition()));
            }
        }
    }
}

File: app/src/main/java/com/liadpaz/amp/adapters/PlaylistAdapter.kt
package com.liadpaz.amp.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.bumptech.glide.Glide
import com.liadpaz.amp.databinding.ItemPlaylistSongBinding
import com.liadpaz.amp.databinding.ItemSongShuffleBinding
import com.liadpaz.amp.interfaces.ItemTouchHelperAdapter
import com.liadpaz.amp.interfaces.OnRecyclerItemClickListener
import com.liadpaz.amp.interfaces.OnStartDragListener
import com.liadpaz.amp.livedatautils.PlaylistsUtil
import com.liadpaz.amp.livedatautils.QueueUtil
import com.liadpaz.amp.utils.Utilities
import com.liadpaz.amp.viewmodels.Song
import java.util.*

class PlaylistAdapter(private val context: Context, private val onMoreClickListener: OnRecyclerItemClickListener, private val onShuffleClickListener: View.OnClickListener, private val itemTouchHelperAdapter: ItemTouchHelperAdapter) : ListAdapter<Song, PlaylistAdapter.SongViewHolder>(object : DiffUtil.ItemCallback<Song>() {
    override fun areItemsTheSame(oldItem: Song, newItem: Song): Boolean {
        return oldItem === newItem
    }

    override fun areContentsTheSame(oldItem: Song, newItem: Song): Boolean {
        return oldItem == newItem
    }
}), ItemTouchHelperAdapter {
    private lateinit var songs: ArrayList<Song>
    private var onStartDragListener: OnStartDragListener? = null
    fun setOnStartDragListener(onStartDragListener: OnStartDragListener) {
        this.onStartDragListener = onStartDragListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SongViewHolder {
        return if (viewType == TYPE_ITEM) {
            SongViewHolder(ItemPlaylistSongBinding.inflate(LayoutInflater.from(context), parent, false), OnRecyclerItemClickListener { _, position: Int ->
                QueueUtil.queue.postValue(ArrayList(songs))
                QueueUtil.queuePosition.postValue(position - 1)
                // TODO: play from queue
            }, onShuffleClickListener)
        } else SongViewHolder(ItemSongShuffleBinding.inflate(LayoutInflater.from(context), parent, false), OnRecyclerItemClickListener { _, _ -> }, onShuffleClickListener)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onBindViewHolder(holder: SongViewHolder, position: Int) {
        if (position != 0) {
            val song = getItem(position - 1)
            val binding = holder.binding as ItemPlaylistSongBinding
            binding.tvSongName.text = song!!.title
            binding.tvSongArtist.text = Utilities.joinArtists(song.artists)
            Glide.with(context).load(song.coverUri).into(binding.ivSongCover)
            binding.btnMore.setOnClickListener { v: View? -> onMoreClickListener.onItemClick(v, position - 1) }
            binding.btnDrag.setOnTouchListener { _, event: MotionEvent ->
                if (event.action == MotionEvent.ACTION_DOWN) {
                    onStartDragListener!!.onStartDrag(holder)
                }
                true
            }
        }
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int) {
        PlaylistsUtil.setIsChanging(true)
        Collections.swap(songs, fromPosition - 1, toPosition - 1)
        itemTouchHelperAdapter.onItemMove(fromPosition - 1, toPosition - 1)
        notifyItemMoved(fromPosition, toPosition)
    }

    override fun onItemDismiss(position: Int) {
        throw UnsupportedOperationException()
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) 0 else TYPE_ITEM
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + 1
    }

    override fun submitList(list: List<Song>?) {
        super.submitList(list?.let { ArrayList(it) }.also { songs = it!! })
    }

    class SongViewHolder(val binding: ViewBinding, onItemClick: OnRecyclerItemClickListener, onShuffleClickListener: View.OnClickListener) : RecyclerView.ViewHolder(binding.root) {

        init {
            if (binding is ItemSongShuffleBinding) {
                binding.root.setOnClickListener(onShuffleClickListener)
            } else {
                itemView.setOnClickListener { v: View? -> onItemClick.onItemClick(v, adapterPosition) }
            }
        }
    }

    companion object {
        private const val TAG = "AmpApp.PlaylistAdapter"
        private const val TYPE_ITEM = 1
    }

}